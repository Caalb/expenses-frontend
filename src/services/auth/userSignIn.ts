import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignInResponse {
  token: string
}

export const userSignIn = async (
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignInResponse>> => {
  const response = await api.post<UserSignInResponse>('/auth/signin', { email, password })

  return response
}
