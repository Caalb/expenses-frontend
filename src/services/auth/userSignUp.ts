import type { AxiosResponse } from 'axios'
import api from '../api'

export interface UserSignUpResponse {
  token: string
}

export const userSignUp = async (
  name: string,
  email: string,
  password: string,
): Promise<AxiosResponse<UserSignUpResponse>> => {
  const response = await api.post<UserSignUpResponse>('/auth/signup', { name, email, password })

  return response
}
