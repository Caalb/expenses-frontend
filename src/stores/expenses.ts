import { createExpense as createExpenseApi } from '@/services/expenses/createExpense'
import { fetchExpenses as fetchExpensesApi } from '@/services/expenses/fetchExpenses'
import type { Expense } from '@/types/expense'
import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useExpensesStore = defineStore('expenses', () => {
  const expenses = ref<Expense[]>([])

  const fetchExpenses = async () => {
    try {
      const response = await fetchExpensesApi()

      expenses.value = response

      return response
    } catch (error) {
      return error
    }
  }

  const createExpense = async (expense: Omit<Expense, 'id'>) => {
    try {
      const response = await createExpenseApi(expense)
      expenses.value = [response.data, ...expenses.value]

      return response
    } catch (error) {
      return error
    }
  }

  return {
    expenses,
    fetchExpenses,
    createExpense,
  }
})
