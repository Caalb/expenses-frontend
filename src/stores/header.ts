import { defineStore } from 'pinia'
import { ref } from 'vue'

export const useHeaderStore = defineStore('header', () => {
  const activeSubmenu = ref<string | null>(null)

  const setActiveSubmenu = (submenu: string) => {
    activeSubmenu.value = activeSubmenu.value === submenu ? null : submenu
  }

  const clearSubmenu = () => {
    activeSubmenu.value = null
  }

  return {
    activeSubmenu,
    setActiveSubmenu,
    clearSubmenu,
  }
})
