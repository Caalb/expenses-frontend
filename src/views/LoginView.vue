<template>
  <QPage class="flex flex-center">
    <QCard class="login-card">
      <QCardSection class="text-center">
        <h5 class="text-h4 q-mt-none q-mb-md text-grey-8">Login</h5>
      </QCardSection>

      <QCardSection>
        <QForm @submit="onSubmit" class="q-gutter-md">
          <QInput
            v-model="formState.email"
            type="email"
            label="E-mail"
            outlined
            :rules="validationRules.email"
          >
            <template v-slot:prepend>
              <QIcon name="email" />
            </template>
          </QInput>

          <QInput
            v-model="formState.password"
            :type="passwordVisible ? 'text' : 'password'"
            label="Senha"
            outlined
            :rules="validationRules.password"
          >
            <template v-slot:prepend>
              <QIcon name="lock" />
            </template>
            <template v-slot:append>
              <QIcon
                :name="passwordVisible ? 'visibility' : 'visibility_off'"
                class="cursor-pointer"
                @click="togglePasswordVisibility"
              />
            </template>
          </QInput>

          <div class="flex justify-between q-mt-sm">
            <QCheckbox v-model="formState.rememberMe" label="Lembrar-me" />
            <QBtn flat color="primary" label="Esqueceu a senha?" @click="handleForgotPassword" />
          </div>

          <div class="full-width q-mt-md">
            <QBtn
              type="submit"
              color="primary"
              label="Entrar"
              class="full-width"
              :loading="loading"
            />
          </div>
        </QForm>
      </QCardSection>

      <QCardSection class="text-center q-pa-sm">
        <p class="text-grey-6">
          Não tem uma conta?
          <QBtn flat color="primary" label="Registre-se" @click="handleRegister" />
        </p>
      </QCardSection>
    </QCard>
  </QPage>
</template>

<script setup lang="ts">
import { ref, reactive } from 'vue'
import { useQuasar } from 'quasar'
import { QPage, QCard, QCardSection, QInput, QCheckbox, QBtn, QIcon, QForm} from 'quasar'

interface LoginCredentials {
  email: string
  password: string
}

type ValidationRule = (val: string) => boolean | string

interface ValidationRules {
  email: ValidationRule[]
  password: ValidationRule[]
}

const $q = useQuasar()

interface FormState extends LoginCredentials {
  rememberMe: boolean
}

const loading = ref<boolean>(false)
const passwordVisible = ref<boolean>(false)

const formState = reactive<FormState>({
  email: '',
  password: '',
  rememberMe: false,
})

const validationRules: ValidationRules = {
  email: [
    (val: string) => !!val || 'E-mail é obrigatório',
    (val: string) => val.includes('@') || 'E-mail inválido',
  ],
  password: [
    (val: string) => !!val || 'Senha é obrigatória',
    (val: string) => val.length >= 6 || 'A senha deve ter pelo menos 6 caracteres',
  ],
}

const togglePasswordVisibility = (): void => {
  passwordVisible.value = !passwordVisible.value
}

const handleForgotPassword = (): void => {
  console.log('Recuperar senha')
}

const handleRegister = (): void => {
  console.log('Ir para página de registro')
}

const onSubmit = async (event: Event): Promise<void> => {
  event.preventDefault()
  try {
    loading.value = true
    await new Promise((resolve) => setTimeout(resolve, 1000))

    console.log('Dados do formulário:', formState)

    $q.notify({
      color: 'positive',
      message: 'Login realizado com sucesso!',
      icon: 'check',
    })
  } catch (error: unknown) {
    $q.notify({
      color: 'negative',
      message: (error as Error)?.message ?? 'Falha ao realizar login',
      icon: 'error',
    })
  }

  loading.value = false
}
</script>

<style scoped>
.login-card {
  width: 100%;
  max-width: 400px;
  padding: 20px;
}
</style>
