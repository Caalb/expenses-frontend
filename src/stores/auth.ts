import { defineStore } from 'pinia'
import { userSignIn } from '../services/auth/userSignIn'
import type { AxiosResponse } from 'axios'
import { userSignUp } from '@/services/auth/userSignUp'

export const useAuthStore = defineStore('auth', () => {
  const handleSignIn = async (
    email: string,
    password: string,
  ): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignIn(email, password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  const handleSignUp = async (data: Record<string, string>): Promise<AxiosResponse | unknown> => {
    try {
      const response = await userSignUp(data.name, data.email, data.password)
      const { token } = response.data

      localStorage.setItem('token', token)
      return response
    } catch (error: unknown) {
      return error
    }
  }

  return {
    handleSignIn,
    handleSignUp,
  }
})
