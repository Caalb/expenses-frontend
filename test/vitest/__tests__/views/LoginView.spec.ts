import router from '@/router'
import { useAuthStore } from '@/stores/auth'
import LoginView from '@/views/LoginView.vue'
import { createTestingPinia } from '@pinia/testing'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'
import { mount, VueWrapper } from '@vue/test-utils'
import { Notify } from 'quasar'
import { getByTestId } from 'test/vitest/__utils__'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import type { ComponentPublicInstance } from 'vue'
import { useRouter } from 'vue-router'

installQuasarPlugin({ plugins: { Notify } })

vi.mock('vue-router')
vi.mock('uuid')

interface LoginViewInstance extends ComponentPublicInstance {
  onSubmit: () => void
}

const createWrapper = () =>
  mount(LoginView, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
        }),
      ],
      mocks: {
        $router: {
          push: vi.fn(),
        },
      },
    },
  }) as VueWrapper<LoginViewInstance>

describe('HeaderActions', () => {
  let wrapper: VueWrapper<LoginViewInstance>
  let store: ReturnType<typeof useAuthStore>

  vi.mocked(useRouter).mockReturnValue({
    ...router,
    push: vi.fn(),
  })

  beforeEach(() => {
    wrapper = createWrapper()
    store = useAuthStore()
  })

  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not submit the login form if the form is invalid', async () => {
    const spy = vi.spyOn(wrapper.vm, 'onSubmit')

    await getByTestId(wrapper, 'login-form').trigger('submit')

    expect(spy).not.toHaveBeenCalled()
  })

  it('should submit the login form', async () => {
    await getByTestId(wrapper, 'email-input').setValue('test@test.com')
    await getByTestId(wrapper, 'password-input').setValue('123456')
    await getByTestId(wrapper, 'login-form').trigger('submit.prevent')

    await wrapper.vm.$nextTick()
    await wrapper.vm.$nextTick()

    expect(store.handleSignIn).toHaveBeenCalledWith('test@test.com', '123456')
  })
})
