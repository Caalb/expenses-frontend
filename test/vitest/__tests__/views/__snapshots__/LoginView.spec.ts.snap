// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeaderActions > should match snapshot 1`] = `
"<main data-v-45f5edd7="" class="q-page flex flex-center" style="min-height: 768px;">
  <div data-v-45f5edd7="" class="q-card login-card">
    <div data-v-45f5edd7="" class="q-card__section q-card__section--vert text-center">
      <h5 data-v-45f5edd7="" class="text-h4 q-mt-none q-mb-md text-grey-8">Login</h5>
    </div>
    <div data-v-45f5edd7="" class="q-card__section q-card__section--vert">
      <form data-v-45f5edd7="" novalidate="" class="q-gutter-md" for="login-form" data-testid="login-form"><label data-v-45f5edd7="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="email">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-45f5edd7="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">email</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="E-mail" name="email" data-testid="email-input" id="email" type="email" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">E-mail</div>
                <!---->
              </div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label><label data-v-45f5edd7="" class="q-field row no-wrap items-start q-field--outlined q-input q-field--labeled q-field--with-bottom full-width" for="password">
          <!---->
          <div class="q-field__inner relative-position col self-stretch">
            <div class="q-field__control relative-position row no-wrap" tabindex="-1">
              <div class="q-field__prepend q-field__marginal row no-wrap items-center"><i data-v-45f5edd7="" class="q-icon notranslate material-icons" aria-hidden="true" role="presentation">lock</i></div>
              <div class="q-field__control-container col relative-position row no-wrap q-anchor--skip"><input class="q-field__native q-placeholder" tabindex="0" aria-label="Senha" name="password" data-testid="password-input" id="password" type="password" value="">
                <div class="q-field__label no-pointer-events absolute ellipsis">Senha</div>
                <!---->
              </div>
              <div class="q-field__append q-field__marginal row no-wrap items-center"><i data-v-45f5edd7="" class="q-icon notranslate material-icons cursor-pointer" aria-hidden="true" role="presentation">visibility_off</i></div>
            </div>
            <div class="q-field__bottom row items-start q-field__bottom--animated">
              <transition-stub name="q-transition--field-message" appear="false" persisted="false" css="true">
                <div class="q-field__messages col"></div>
              </transition-stub>
              <!---->
            </div>
          </div>
          <!---->
        </label>
        <div data-v-45f5edd7="" class="full-width q-mt-md"><button data-v-45f5edd7="" class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable full-width" tabindex="0" type="submit"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Entrar</span></span>
            <transition-stub name="q-transition--fade" appear="false" persisted="false" css="true">
              <!---->
            </transition-stub>
          </button></div>
      </form>
    </div>
    <div data-v-45f5edd7="" class="q-card__section q-card__section--vert text-center q-pa-sm">
      <p data-v-45f5edd7="" class="text-grey-6"> NÃ£o tem uma conta? <button data-v-45f5edd7="" class="q-btn q-btn-item non-selectable no-outline q-btn--flat q-btn--rectangle text-primary q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><span class="block">Registre-se</span></span></button></p>
    </div>
  </div>
</main>"
`;
